### PR TITLE
Use real filesystem in ops.testing backend

### PR DESCRIPTION
**Pull request by weiiwang01**
This pull request refactors the testing module of the Ops library, switching its backend from an in-memory virtual filesystem to a real filesystem for simulating filesystem operations within containers. Furthermore, it exposes a new API, `Harness.get_filesystem_root`, enabling developers to interact with the simulated filesystem during test cases for both populating and inspecting the testing container filesystem.

Supersede https://github.com/canonical/operator/pull/933

## Changelog

 - Refactored the testing module to use a real filesystem for simulating container filesystem operations.
 - Introduced a new API, `ops.testing.Harness.get_filesystem_root`, for accessing and manipulating the simulated filesystem during testing.

### CODE REVIEW

File: ops/testing.py
Comment (benhoyt) [inline ops/testing.py:2564]: In the `str` and `bytes` cases we already have the content in memory, but I think we should avoid reading the whole thing into memory if it's a file/buffer. Can we instead use `copyfileobj`, which only reads in 64K chunks, and do something like this (untested):

```python
if isinstance(source, str):
    file_path.write_text(source, encoding=encoding)
elif isinstance(source, bytes):
    file_path.write_bytes(source)
else:
    with file_path.open('w' if encoding is None else 'wb', encoding=encoding) as f:
        shutil.copyfileobj(source, f)
```

File: ops/testing.py
Comment (benhoyt) [inline ops/testing.py:2572]: I don't understand the use of `secrets.token_urlsafe` here. Can you explain? Also, when would the `NotADirectoryError` happen here?

File: ops/testing.py
Comment (benhoyt) [inline ops/testing.py:2604]: Would this work?

```suggestion
        for file_info in file_infos:
            rel_path = os.path.relpath(file_info.path, start=self._root)
            rel_path = '/' if rel_path == '.' else '/' + rel_path
        return file_infos
```

File: ops/testing.py
Comment (benhoyt) [inline ops/testing.py:2635]: Similar to above, should we just skip the `os.chmod()` if permissions is None?

File: ops/testing.py
Comment (benhoyt) [inline ops/testing.py:2656]: As much as I like not nesting, I think this logic would be a bit clearer with an if/else and only one `file_path.is_dir()`:

```python
if file_path.is_dir():
    if recursive:
        shutil.rmtree(file_path)
    else:
        try:
            file_path.rmdir()
        except OSError:
            ...
else:
    file_path.unlink()
```

File: test/test_testing.py
Comment (benhoyt) [inline test/test_testing.py:2692]: This is a good test as far as it goes, but I think we need a few more test cases to test that 1) files that the test writes via `get_filesystem_root` can then be pulled and listed, and 2) files pushed and make_dir'd can then be accessed via `get_filesystem_root`. Or am I missing something?

File: .github/workflows/framework-tests.yaml
Comment (benhoyt) [inline .github/workflows/framework-tests.yaml:41]: I don't love how this shows up in the test summary in GitHub, e.g., "ubuntu-latest, 3.10, false" -- it's unclear what "false" means. Also, I'd rather not run all the tests twice, and I don't think we need to or should be running all the tests as root anyway.

Can we use a suitable test name prefix, and then do something like we do with the RealPebble tests below? Perhaps a TestFilesystemAsRoot class and then use `-k FilesystemAsRoot` filter. I think having a separate job is cleaner (even though it's a bit more duplication).

**Review by benhoyt — CHANGES_REQUESTED** [review]
Thanks for your continued effort here. This is really good work. I've left a bunch of comments. Mostly minor, but one discussion point about whether we should do the user/group stuff at all. I'll also try to schedule a meeting tomorrow to finish discussion on this.

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py:2631]: In that case the test result will affected by the umask in developer's environment, which I would like to avoid.

File: test/test_testing.py
Comment (weiiwang01) [reply test/test_testing.py:2692]: Yes, we already have a suite of tests for the filesystem, using the Pebble APIs (blackbox tests). I think I will add some more tests related to the `get_filesystem_root` method, since that's an exposed API now. Thanks!

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py:2523]: That actually complicates the situation since `uid` and `gid` are both required by `os.chown` and they must be integer, not `None`.

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py:2544]: Indeed, it's better. Fixed, thanks!

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py:2465]: Removed, thanks!

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py:1981]: It does look like a duplicate, removed.

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py:1894]: Removed, thanks!

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py:2652]: Yes, updated!

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py:2571]: When pebble transferring a file, it will first write to a temporary path (filename + random string + ~), and rename to the final destination when the transferring is finished. This is just mimicking the pebble error, while  `secrets.token_urlsafe` is just a quick way to generate a random string.

This error will happen if the parent directory doesn't exist and `make_dirs` is set to `False`.

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py:2563]: It's better, updated, thanks!

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py:2550]: Answered down below, thanks!

**Comment by weiiwang01** [comment]
> Thanks for your continued effort here. This is really good work. I've left a bunch of comments. Mostly minor, but one discussion point about whether we should do the user/group stuff at all. I'll also try to schedule a meeting tomorrow to finish discussion on this.

Thanks for the fantastic review! Yes, let's talk about the permission issue.

Another topic I want to discuss is about simulating storage mounts. Currently, we are using symbolic links to simulate storage mounts. However, actual storage mounts have some special properties that symbolic links can't simulate. For example, when a filesystem is attached to a directory, the original content of that directory becomes "shadowed". Additionally, it's an error to delete the mounting directory. Also, `list_files` should display the mounting directory as a directory, but right now, it shows as a symbolic link.

Of course, we can simulate these behaviors in the harness, but I feel like this introduces dishonesty into the design. I would prefer to let the user know that we are using symbolic links to simulate storage mounts and it's the users' responsibility to avoid tests involves unique properties of mounting directories.

File: .github/workflows/framework-tests.yaml
Comment (weiiwang01) [reply .github/workflows/framework-tests.yaml:41]: I think the tests depend on the outcome of the discussion, I will defer the changes for now.

**Comment by benhoyt** [comment]
Per yesterday's discussion, @weiiwang01's going to simplify by removing the user/group handling, and (for now at least) document where the test harness isn't a perfect match. As a consequence, we can always remove the tests running as root (we don't want to run tests as root or encourage charmers to do that). He's also going to remove the `secrets.token_urlsafe` call and just leave that out -- people shouldn't be testing exact contents of error messages anyway.

File: ops/testing.py
Comment (benhoyt) [inline ops/testing.py:556]: IMO we should drop the sentence "Users should be cognizant of..." (here and below). We include the first sentence exactly so users are cognizant of it.

File: ops/testing.py
Comment (benhoyt) [inline ops/testing.py:1472]: I'd prefer to avoid the `>>> ` prefix on each line, as this isn't typically something you'd type into the Python REPL. To do this, use a double-colon on `Example usage::` and then make the follow code block indented by four spaces:

```suggestion
        Example usage::

            # charm.py
            import ops
            class ExampleCharm(ops.CharmBase):
                def __init__(self, *args):
                    super().__init__(*args)
                    self.hostname = open("/etc/hostname").read()

            # test_charm.py
            from ops.testing import Harness
            harness = Harness(ExampleCharm)
            root = harness.get_filesystem_root("mycontainer")
            (root / "etc" / "hostname").write_text("hostname.example.com")
            harness.begin()
```

File: ops/testing.py
Comment (benhoyt) [inline ops/testing.py:2571]: Why isn't `file_info.name` correct already?

File: ops/testing.py
Comment (benhoyt) [inline ops/testing.py:2620]: Given there's an early return above, I prefer to simplify and avoid the else (similar to how the structure was before).

File: ops/testing.py
Comment (benhoyt) [inline ops/testing.py:2629]: No need for this `return`. Each of the branches performs an operation, then the method implicitly returns at the end.

File: test/test_testing.py
Comment (benhoyt) [inline test/test_testing.py:4394]: Yeah, if that's the case, that seems like a bug. Can you please open a separate canonical/operator issue for that?

**Review by benhoyt — APPROVED** [review]
I think this is great. Just a couple of nit comments. And also get a second review (probably John's as we've discussed this with him).

**Comment by weiiwang01** [comment]
> Per yesterday's discussion, @weiiwang01's going to simplify by removing the user/group handling, and (for now at least) document where the test harness isn't a perfect match. As a consequence, we can always remove the tests running as root (we don't want to run tests as root or encourage charmers to do that). He's also going to remove the `secrets.token_urlsafe` call and just leave that out -- people shouldn't be testing exact contents of error messages anyway.

I've eliminated the ownership handling and restructured the code according to the recommendation. However, while writing the tests, I came across an problem: the `pull_path` method within `ops.model.Container` doesn't retrieve empty directories. I'm curious to know if this is by design, similar to how git operates.

**Comment by benhoyt** [comment]
Hi @weiiwang01 Thanks -- yep, I replied to your comments above. I've opened an issue on the `pull_path` empty directory issue: https://github.com/canonical/operator/issues/968 -- also see the other comments above.

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py:1482]: Added a comment here as suggested.

File: test/test_testing.py
Comment (weiiwang01) [reply test/test_testing.py:2692]: I have added more unit tests for the `get_filesystem_root` method, thanks!

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py]: Removed, thanks!

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py]: Removed, thanks!

File: ops/testing.py
Comment (jameinel) [reply ops/testing.py:2572]: if it is only for the root, why set it always, vs just:
```
if rel_path == "/":
  // override the root name, since it should be considered the local root, without any filename
  file_info.name = "/"
  ```
 
It seems clearer than always setting it to the value it should already have. (and avoids computing basename() for every file in every directory if you are iterating over them)

**Review by jameinel — APPROVED** [review]

File: ops/testing.py
Comment (benhoyt) [reply ops/testing.py:1662]: Yeah, I thought the same too. But then I did a timing test, and on my laptop I can `os.mkdir()` 10,000 directories in 0.1s and 100,000 in 1.6s (and a bit less than that for `os.rmdir()`). So 3 dir creates + deletes adds about 50 microseconds to each test. Linux FS operations are quick!

File: ops/testing.py
Comment (benhoyt) [reply ops/testing.py:2572]: John's suggestion looks clearer to me too.

**Comment by benhoyt** [comment]
Thanks. Let's just make that `file_info.name = "/"` tweak that John mentioned, and then I'll see about merging this!

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py:1662]: We can prefer `/dev/shm` as the `tempdir` to expedite the filesystem calls. But as the benchmark shows, it doesn't really have a significant influence on the overall performances.

**Comment by benhoyt** [comment]
Yeeha! Thanks for this @weiiwang01. This will be included with the next version of ops, 2.5.0, planned for release late July / early August.

**Pull request merged**
